### PR TITLE
mgr/rook: refactor rook apply/add nfs

### PR DESCRIPTION
**Pull request by varshar16**
Fixes: https://tracker.ceph.com/issues/47923

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by jtlayton — COMMENTED** [review]
This looks like a reasonable change to me, but I'm not as heavily involved with this code these days, so someone else should approve it.

File: doc/mgr/orchestrator.rst [hunk 1]
```diff
@@ -757,7 +757,7 @@ This is an overview of the current implementation status of the orchestrators.
  device ls                           ✔      ✔
  iscsi add                           ⚪     ✔
  mds add                             ✔      ✔
- nfs add                             ✔      ✔
+ nfs add                             ⚪      ✔
```
Comment (travisn) [inline doc/mgr/orchestrator.rst:760]: Isn't this PR fixing the nfs add scenario? Seems like it should still be checked?

Comment (travisn) [reply doc/mgr/orchestrator.rst:760]: Ok, I see now the `apply` is preferred

**Review by travisn — APPROVED** [review]

File: src/pybind/mgr/rook/rook_cluster.py [hunk 2]
```diff
@@ -389,8 +389,8 @@ def apply_filesystem(self, spec):
         # TODO use spec.placement
         # TODO warn if spec.extended has entries we don't kow how
         #      to action.
-        def _update_fs(current, new):
-            # type: (cfs.CephFilesystem, cfs.CephFilesystem) -> cfs.CephFilesystem
+        def _update_fs(new):
+            # type: (cfs.CephFilesystem) -> cfs.CephFilesystem
```
Comment (mgfritch) [inline src/pybind/mgr/rook/rook_cluster.py:393]:
```suggestion
        def _update_fs(new: cfs.CephFilesystem) -> cfs.CephFilesystem:
```

converting these to py3 style ?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 3]
```diff
@@ -447,40 +447,50 @@ def _create_zone():
...
             new.spec.gateway.instances = spec.placement.count or 1
             return new
 
         return self._create_or_patch(
             cos.CephObjectStore, 'cephobjectstores', name,
             _update_zone, _create_zone)
 
-    def add_nfsgw(self, spec):
+    def apply_nfsgw(self, spec):
+        # type: (ServiceSpec) -> None
```
Comment (mgfritch) [inline src/pybind/mgr/rook/rook_cluster.py:459]: For the make check failures:

```suggestion
    def apply_nfsgw(self, spec: ServiceSpec):
        # type: (ServiceSpec) -> None
```

(also need to add import for `NFSServiceSpec` type)

File: src/pybind/mgr/rook/rook_cluster.py [hunk 4]
```diff
@@ -447,40 +447,50 @@ def _create_zone():
...
         # TODO use spec.placement
         # TODO warn if spec.extended has entries we don't kow how
         #      to action.
+        cnt = spec.placement.count or 1
```
Comment (mgfritch) [inline src/pybind/mgr/rook/rook_cluster.py:463]: should we ensure the `PlacementSpec` always has either a count or a list of hosts?

File: src/pybind/mgr/rook/module.py [hunk 5]
```diff
@@ -413,11 +413,6 @@ def _service_add_decorate(self, typename, spec, func):
             mgr=self
         )
 
-    def add_nfs(self, spec):
-        # type: (NFSServiceSpec) -> RookCompletion
-        return self._service_add_decorate("NFS", spec,
-                                          self.rook_cluster.add_nfsgw)
-
```
Comment (mgfritch) [inline src/pybind/mgr/rook/module.py:420]: Do we want to remove these or try to fix them per #37771?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 2 above)
Comment (varshar16) [reply src/pybind/mgr/rook/rook_cluster.py]: Done

File: src/pybind/mgr/rook/rook_cluster.py (hunk 3 above)
Comment (varshar16) [reply src/pybind/mgr/rook/rook_cluster.py]: Done

File: src/pybind/mgr/rook/module.py (hunk 5 above)
Comment (varshar16) [reply src/pybind/mgr/rook/module.py:420]: > Do we want to try to fix these per #37771?

I have closed that PR.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 4 above)
Comment (varshar16) [reply src/pybind/mgr/rook/rook_cluster.py:458]: I would prefer not changing it in this PR. As validation of hosts needs to be done in other places too.

Comment (mgfritch) [reply src/pybind/mgr/rook/rook_cluster.py:458]: mind updating the TODO comment to remind us to add pod affinities based on the list of hosts in the PlacementSpec?

**Review by mgfritch — APPROVED** [review]
minor nit, but otherwise lgtm!

File: src/pybind/mgr/rook/rook_cluster.py (hunk 4 above)
Comment (varshar16) [reply src/pybind/mgr/rook/rook_cluster.py]: Done!

**Comment by varshar16** [comment]
Rebased

**Comment by varshar16** [comment]
@batrick Builds successfully: https://shaman.ceph.com/repos/ceph/wip-refactor-rook-apply-add-nfs/a251cea7144d4a1f1cb52c6f49317606c3932e42/

It fails to build for centos-8 with crimson flavour. The failure is not due to this PR. https://shaman.ceph.com/builds/ceph/wip-refactor-rook-apply-add-nfs/a251cea7144d4a1f1cb52c6f49317606c3932e42/crimson/237112/

**Comment by batrick** [comment]
Note for others: merging this because there are no tests for the mgr/rook module yet except what Varsha is developing.

**Pull request merged**